### PR TITLE
Spark: vectorized/non-vectorized read compound constant type exception (#3139)

### PR DESCRIPTION
**Pull request by rice668**
This patch fixes a bug in Spark reading ORC compound type constants.

This bug can be reproduced by modifying test TestPartitionValues::testPartitionedByNestedString, to let it write with ORC format instead of parquet, e.g. with the following change:

    // write into iceberg
    sourceDF.write()
        .format("iceberg")
        .option(WRITE_FORMAT, format) // add this line
        .mode(SaveMode.Append)
        .save(baseLocation);
And the test would fail with:

```
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.get(ArrayList.java:435)
	at org.apache.iceberg.orc.OrcValueReaders$StructReader.<init>(OrcValueReaders.java:161)
	at org.apache.iceberg.spark.data.SparkOrcValueReaders$StructReader.<init>(SparkOrcValueReaders.java:143)
	at org.apache.iceberg.spark.data.SparkOrcValueReaders.struct(SparkOrcValueReaders.java:70)
	at org.apache.iceberg.spark.data.SparkOrcReader$ReadBuilder.record(SparkOrcReader.java:75)
	at org.apache.iceberg.spark.data.SparkOrcReader$ReadBuilder.record(SparkOrcReader.java:65)
	at org.apache.iceberg.orc.OrcSchemaWithTypeVisitor.visitRecord(OrcSchemaWithTypeVisitor.java:71)
	at org.apache.iceberg.orc.OrcSchemaWithTypeVisitor.visit(OrcSchemaWithTypeVisitor.java:38)
	at org.apache.iceberg.orc.OrcSchemaWithTypeVisitor.visit(OrcSchemaWithTypeVisitor.java:32)
	at org.apache.iceberg.spark.data.SparkOrcReader.<init>(SparkOrcReader.java:52)
	at org.apache.iceberg.spark.source.RowDataReader.lambda$newOrcIterable$2(RowDataReader.java:164)
	at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:108)
	at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:45)
	at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
	at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:73)
	at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:78)
```
In ```RowDataReader::newOrcIterable```, we exclude the struct field and create an OrcIterable with an empty schema, because the inner string is a constant (partition value). And later on we hit the exception when constructing the readers.

### CODE REVIEW

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 1]
```diff
@@ -59,6 +66,80 @@
...
+    List<Types.NestedField> columns = tableSchema.columns();
+
+    for (Types.NestedField nestedField: columns) {
+      compoundTypeConstants = visitCompoundTypesByDFS(nestedField, idToConstant);
+    }
+    compoundTypeConstants.putAll(idToConstant);
+    return compoundTypeConstants;
+  }
+
+  protected boolean containsCompoundType(Schema tableSchema) {
```
Comment (lirui-apache) [inline spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:80]: I don't think we need this. `visitCompoundTypesByDFS` skips primitive types anyway.

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 2]
```diff
@@ -59,6 +66,80 @@
...
+        for (Types.NestedField childField :childFieldsInStruct) {
+          visitCompoundTypesByDFS(childField, idToConstant);
+          childFieldID.add(childField.fieldId());
+        }
+        if (idToConstant.keySet().containsAll(childFieldID)) {
+          Object[] objects = new Object[childFieldID.size()];
+          for (int i =0; i < objects.length; i++) {
+            objects[i] = idToConstant.get(childFieldID.get(i));
+          }
+          moreConstants.put(nestedField.fieldId(), new GenericInternalRow(objects));
```
Comment (lirui-apache) [inline spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:106]: Why not just reuse `idToConstant`

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 3]
```diff
@@ -59,6 +66,80 @@
...
+        }
+        break;
+      case MAP:
+        int keyId = nestedField.type().asMapType().keyId();
+        int valueId = nestedField.type().asMapType().valueId();
+        visitCompoundTypesByDFS(nestedField, idToConstant);
```
Comment (lirui-apache) [inline spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:112]: Will this lead to infinite recursion?

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 4]
```diff
@@ -59,6 +66,80 @@
     this.batchSize = size;
   }
 
+  protected Map<Integer, ?> findMoreCompoundConstants(Schema tableSchema, Map<Integer, ?> idToConstant) {
+    Map<Integer, Object> compoundTypeConstants = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (zhangminglei) [reply spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:70]: actually, we can use the original map instead.

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java (hunk 1 above)
Comment (zhangminglei) [reply spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:80]: you are right. i add this for a easy way to view. will remove this soon.

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java (hunk 2 above)
Comment (zhangminglei) [reply spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:106]: will change. thanks.

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java (hunk 3 above)
Comment (zhangminglei) [reply spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java]: Fixed done.

File: spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java [hunk 5]
```diff
@@ -78,6 +89,81 @@
...
+    List<Types.NestedField> columns = tableSchema.columns();
+    for(Types.NestedField nestedField: columns) {
+      if (nestedField.type().isStructType() || nestedField.type().isMapType() || nestedField.type().isListType()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  protected Map<Integer, ?> findMoreCompoundConstants(Schema tableSchema, Map<Integer, ?> idToConstant) {
```
[the remark this answers is not in the document]
Comment (zhangminglei) [reply spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:87]: will change.

**Comment by rdblue** [comment]
Running CI. Can you update the summary/description to note that this fixes a bug with ORC?

**Comment by zhangminglei** [comment]
Hi @rdblue , Could you please take a look on this ?

**Comment by rdblue** [comment]
Running CI.

**Comment by zhangminglei** [comment]
@rdblue I will port these changes to spark2.4/3.0/3.2 respectively.

**Comment by zhangminglei** [comment]
Hi, @rdblue  Could you please review that again ? Thanks.

File: spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java [hunk 6]
```diff
@@ -402,14 +402,13 @@ public void testPartitionedByNestedString() throws Exception {
...
-            ),
-            false, Metadata.empty()
-        )
+        new StructField("col1",
+          DataTypes.createStructType(
+            new StructField[] {
+                new StructField("f1", DataTypes.StringType, false, Metadata.empty())
+            }
+                    ), false, Metadata.empty()
+            )
```
Comment (rdblue) [inline spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:411]: Nearly all of the changes in this file are to formatting or names. Can you remove any changes that are not strictly necessary for what you're trying to fix?

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 7]
```diff
@@ -196,4 +198,40 @@ protected static Object convertConstant(Type type, Object value) {
...
+          Map<Integer, Object> idToConstant) {
+    switch (nestedField.type().typeId()) {
+      case STRUCT:
+        List<Types.NestedField> childFieldsInStruct = nestedField.type().asStructType().fields();
+        List<Integer> childFieldID = new ArrayList<>();
+        for (Types.NestedField childField : childFieldsInStruct) {
+          visitCompoundTypesByDFS(childField, idToConstant);
+          childFieldID.add(childField.fieldId());
+        }
+        if (idToConstant.keySet().containsAll(childFieldID)) {
```
Comment (rdblue) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:221]: Looks like this is creating any struct that it can when all of the struct's children are present in the constants map.

I think this would need to be refactored so that the method is clear about what it is doing. Simply "visit" doesn't really describe what is going on here. It is updating the constants map.

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 8]
```diff
@@ -196,4 +198,40 @@ protected static Object convertConstant(Type type, Object value) {
...
+    List<Types.NestedField> columns = tableSchema.columns();
+
+    for (Types.NestedField nestedField : columns) {
+      visitCompoundTypesByDFS(nestedField, (Map<Integer, Object>) idToConstant);
+    }
+  }
+
+  protected void visitCompoundTypesByDFS(
+          Types.NestedField nestedField,
+          Map<Integer, Object> idToConstant) {
```
Comment (rdblue) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:212]: I'm not sure that this should alter the incoming map. It would be better to build a new constants map with the extra fields.

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 9]
```diff
@@ -196,4 +198,40 @@ protected static Object convertConstant(Type type, Object value) {
...
+    switch (nestedField.type().typeId()) {
+      case STRUCT:
```
Comment (rdblue) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:214]: In Iceberg, we separate schema traversal logic from other logic using visitors. I think that you should use a type visitor to implement this rather than traversing a type tree and doing some task in the same recursive function.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 10]
```diff
@@ -196,4 +198,39 @@ protected static Object convertConstant(Type type, Object value) {
     }
     return value;
   }
+
+  protected void findMoreCompoundConstants(Schema tableSchema, Map<Integer, ?> idToConstant) {
```
Comment (rdblue) [inline spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:202]: Can you please update just one version of Spark and then we'll port the changes to the other versions in separate PRs? That way, each PR is smaller and we don't have to review identical code multiple times.

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java [hunk 11]
```diff
@@ -119,6 +122,8 @@ public UTF8String getUTF8String(int rowId) {
 
   @Override
   protected ColumnVector getChild(int ordinal) {
-    throw new UnsupportedOperationException("ConstantColumnVector only supports primitives");
+    DataType sparkType = ((StructType) type).fields()[ordinal].dataType();
+    Type childType = SparkSchemaUtil.convert(sparkType);
+    return new ConstantColumnVector(childType, batchSize, ((InternalRow) constant).get(ordinal, sparkType));
```
Comment (rdblue) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java:127]: How are these changes related to ORC?

File: spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java (hunk 6 above)
Comment (zhangminglei) [reply spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:411]: Yes. I will change.

File: spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 10 above)
Comment (zhangminglei) [reply spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:202]: I will only make this pr to spark3.2

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java (hunk 11 above)
Comment (zhangminglei) [reply spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java:127]: I think the code can be removed for now, because previous the ORC constant reading generates a StructReader that needs to be retrieved from the child node, there should be no such problem now.

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 9 above)
Comment (zhangminglei) [reply spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:214]: This is a good idea for refactoring, thanks.

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java (hunk 11 above)
Comment (zhangminglei) [reply spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java:127]:
```
ColumnarRow #

  @Override
  public UTF8String getUTF8String(int ordinal) {
    return data.getChild(ordinal).getUTF8String(rowId);
  }
```
To get a column value in a row use `getChild` when performing spark vectorization reads to fetch each column element.

image

**Comment by zhangminglei** [comment]
@rdblue I updated the code again. Could you please take a look when you have a chance ?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java [hunk 12]
```diff
@@ -392,8 +393,7 @@ public void testPartitionedByNestedString() throws Exception {
     // schema and partition spec
     Schema nestedSchema = new Schema(
         Types.NestedField.required(1, "struct",
-            Types.StructType.of(Types.NestedField.required(2, "string", Types.StringType.get()))
-        )
+            Types.StructType.of(Types.NestedField.required(2, "string", Types.StringType.get())))
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:396]: Please revert the non-functional changes in this file.

Comment (zhangminglei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java]: Sorry, I miss the code specification, now it has been updated, Could you have a look again ? Thank you.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java [hunk 13]
```diff
@@ -119,6 +122,8 @@ public UTF8String getUTF8String(int rowId) {
 
   @Override
   public ColumnVector getChild(int ordinal) {
-    throw new UnsupportedOperationException("ConstantColumnVector only supports primitives");
+    DataType sparkType = ((StructType) type).fields()[ordinal].dataType();
+    Type childType = SparkSchemaUtil.convert(sparkType);
+    return new ConstantColumnVector(childType, batchSize, ((InternalRow) constant).get(ordinal, sparkType));
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java:127]: It looks like the approach in this PR is to use existing constant handling to fix nested column construction. I'm not sure that this is the right approach. There isn't anything specific to ORC here, so I'm not convinced that the ORC readers handle nested constant values correctly. Avoiding a problem is an okay solution as long as we're confident that it is always avoided.

At a minimum, I think we should have tests for partitioning by nested fields in structs that are not constants and would not get created by `findMoreCompoundConstants`.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java (hunk 12 above)
Comment (rdblue) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java]: Looks like this is still a problem.

Comment (zhangminglei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java]: Please see the latest code ~

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java (hunk 13 above)
Comment (zhangminglei) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ConstantColumnVector.java:127]: > There isn't anything specific to ORC here,so I'm not convinced that the ORC readers handle nested constant values correctly

Ans: The change itself is not specific to ORC here, but ORC is currently used that way, and Parquet's vectorization read do not use this, and this is a general practice, not to ORC itself.

>  I'm not sure that this is the right approach.

Ans:  When doing the vectorization read for orc, ```VectorizedSparkOrcReaders#StructConverter```use ```idToConstant``` and will convert constant values to ```ConstantColumnVector```, but```ConstantColumnVector``` cannot represent a Struct Type currently. So I need a struct constant.

According to Spark's API, a ```ColumnVector``` that represents a struct must implement the ```getChild``` method.
https://github.com/apache/spark/blob/v3.2.0/sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java#L219
https://github.com/apache/spark/blob/v3.2.0/sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java#L302

Since ```getChild``` returns a ```ColumnVector```, and each child in a struct type is a constant also, so it can return a ```ConstantColumnVector``` I think.

Besides , I will add more tests.

**Comment by zhangminglei** [comment]
@rdblue Could you take a look again when you have a chance ?
